core/src/apps/webauthn/__init__.py

```diff
@@ -749,3 +1230,413 @@ def msg_error(cid: int, code: int) -> Cmd:
+    dig = hashlib.sha256()
+    dig.update(authenticator_data)
+    dig.update(client_data_hash)
+    sig = nist256p1.sign(privkey, dig.digest(), False)
+    sig = der.encode_seq((sig[1:33], sig[33:]))
+
+    # Encode the authenticatorMakeCredential response data.
+    return cbor.encode(
+        {
+            _MAKECRED_RESP_FMT: "packed",
+            _MAKECRED_RESP_AUTH_DATA: authenticator_data,
+            _MAKECRED_RESP_ATT_STMT: {"alg": _COSE_ALG_ES256, "sig": sig},
+        }
+    )
+
+
+def cbor_get_assertion(req: Cmd, dialog_mgr: DialogManager) -> Optional[Cmd]:
+    if not storage.is_initialized():
+        if __debug__:
+            log.warning(__name__, "not initialized")
+        return cbor_error(req.cid, _ERR_OPERATION_DENIED)
+
+    try:
+        param = cbor.decode(req.data[1:])
+        rp_id = param[_GETASSERT_CMD_RP_ID]
+        rp_id_hash = hashlib.sha256(rp_id).digest()
+
+        allow_list = param.get(_GETASSERT_CMD_ALLOW_LIST, [])
+        if allow_list:
+            # Get all credentials from the allow list that belong to this authenticator.
+            cred_list = []
+            for credential_descriptor in allow_list:
+                if credential_descriptor["type"] != "public-key":
+                    continue
+                cred = Credential.from_bytes(credential_descriptor["id"], rp_id_hash)
+                if cred is not None:
+                    if cred.rp_id is None:
+                        cred.rp_id = rp_id
+                    cred_list.append(cred)
+        else:
+            # Allow list is empty. Get resident credentials.
+            if not _ALLOW_RESIDENT_CREDENTIALS:
+                return cbor_error(req.cid, _ERR_UNSUPPORTED_OPTION)
+            cred_list = get_resident_credentials(rp_id_hash)
+
+        # Sort credentials by time of creation.
+        cred_list.sort()
+
+        # Check that the pinAuth parameter is absent. Client PIN is not supported.
+        if _GETASSERT_CMD_PIN_AUTH in param:
+            return cbor_error(req.cid, _ERR_PIN_AUTH_INVALID)
+
+        # Get options.
+        options = param.get(_GETASSERT_CMD_OPTIONS, {})
+        user_presence = options.get("up", True)
+        user_verification = options.get("uv", False)
+
+        # Get extensions.
+        hmac_secret = param.get(_GETASSERT_CMD_EXTENSIONS, {}).get("hmac-secret", None)
+
+        client_data_hash = param[_GETASSERT_CMD_CLIENT_DATA_HASH]
+    except Exception:
+        return cbor_error(req.cid, _ERR_INVALID_CBOR)
+
+    # Check data types.
+    if (
+        not isinstance(hmac_secret, (dict, type(None)))
+        or not isinstance(client_data_hash, (bytes, bytearray))
+        or not isinstance(user_presence, bool)
+        or not isinstance(user_verification, bool)
+    ):
+        return cbor_error(req.cid, _ERR_INVALID_CBOR)
+
+    # User verification cannot happen without user presence.
+    user_presence = user_presence or user_verification
```

Comment:
I interpret CTAP2 in the following way:

| user_presence | user_verification | PIN needed | User interaction needed |
|---------------|-------------------|------------|-------------------------|
| False         | False             | No         | No                      |
| False         | True              | Yes        | No                      |
| True          | False             | No         | Yes                     |
| True          | True              | Yes        | Yes                     |

Reply:
ACK